### PR TITLE
First step to rebuild Parser.

### PR DESCRIPTION
**Pull request by ice1000**
Create pull request for reviewing.

### CODE REVIEW

File: src/org/lice/parse2/Lexer.java [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+public class Lexer {
+	private static final String upperCaseLetters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+	private static final String lowerCaseLetters = "abcdefghijklmnopqrstuvwxyz";
+	private static final String commonSymbols = "!@#$%^&*_=:<>.?/\\+-*/%[]{}|";
+	private static final String binNumbers = "01";
+	private static final String octNumbers = "01234567";
+	private static final String decNumbers = "0123456789";
+	private static final String hexNumbers = "ABCDEF";
+	private static final String blanks = " \f\n\t,";
```
Comment (ice1000) [inline src/org/lice/parse2/Lexer.java]: Missing `\r`

File: src/org/lice/parse2/Lexer.java [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+	public Token peekOneToken() {
+		assert currentTokenIndex+1 < this.tokenBuffer.size();
+		return this.tokenBuffer.get(this.currentTokenIndex+1);
+	}
+
+	public void nextToken() {
+		this.currentTokenIndex++;
+	}
+
+	@Deprecated
```
Comment (ice1000) [inline src/org/lice/parse2/Lexer.java]: Missing reason and replacement

File: src/org/lice/parse2/Lexer.java [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+		this.currentTokenIndex = 0;
+	}
+
+	private void lexIdentifier() {
+		int line = this.line;
+		int startAtCol = this.col;
+		String str = scanFullString(idChars);
+
+		assert this.col > startAtCol;
+		assert this.line == line;
```
Comment (ice1000) [inline src/org/lice/parse2/Lexer.java]: Use tests instead of asserts.

**Review by ice1000 — COMMENTED** [review]
After fixing all thes we can start converting this parser into Kotlin.

File: test/org/lice/parse2/LexNumberTest.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+import static org.junit.Assert.assertTrue;
+
+public class LexNumberTest {
+	@Test(timeout = 100)
+	public void testLexIdentifier() {
+		String srcCode = "0x74138HC";
+		Lexer l = new Lexer(srcCode);
+		assertTrue(l.currentToken().getKind() == Token.TokenKind.Identifier);
+		assertTrue(l.peekOneToken().getKind() == Token.TokenKind.EOI);
```
Comment (ice1000) [inline test/org/lice/parse2/LexNumberTest.java]: Good!

**Comment by chuigda** [comment]
The recent two commits unified `LexException` and `ParseException`, added solutions for comments, finalized solutions for numbers and renamed TokenKind with TokenType

File: src/org/lice/parse2/Lexer.java [hunk 5]
```diff
@@ -2,24 +2,23 @@
...
-	private static final String commonSymbols = "!@#$%^&*_=:<>.?/\\+-*/%[]{}|";
-	private static final String binNumbers = "01";
-	private static final String octNumbers = "01234567";
-	private static final String decNumbers = "0123456789";
-	private static final String hexNumbers = "0123456789ABCDEF";
+	private static final String commonSymbols = "!@#$%^&*_=:<>.?/\\+-*%[]{}|";
+	private static final String binDigits = "01";
+	private static final String octDigits = "01234567";
+	private static final String decDigits = "0123456789";
+	private static final String hexDigits = "0123456789ABCDEF";
```
Comment (ice1000) [inline src/org/lice/parse2/Lexer.java]: Remember to convert it into `Range` when start migrating to Kotlin.

File: src/org/lice/parse2/Token.java [hunk 6]
```diff
@@ -3,25 +3,51 @@
...
+		HexNumber,
+		LongInteger,
+		BigInteger,
+		FloatNumber,
+		DoubleNumber,
 		StringLiteral,
 		Identifier,
 		LispKwd,
 		EOI
 	}
```
Comment (ice1000) [inline src/org/lice/parse2/Token.java]: Missing `BigDecimal`.

Comment (ice1000) [reply src/org/lice/parse2/Token.java]: Recommended to use `BigInt` and `BigDec`, to avoid naming conflict.

File: src/org/lice/util/Error.kt [hunk 7]
```diff
@@ -12,11 +12,36 @@ package org.lice.util
...
+			}
+		System.err.println();
+	}
+}
 
 class InterpretException(string: String) : RuntimeException(string) {
 	constructor(string: String, meta: MetaData) : this("$string\nat line: ${meta.lineNumber}")
 
+	fun prettyPrint(cachedCodeLines: Array<String>) {
+	}
```
Comment (ice1000) [inline src/org/lice/util/Error.kt]: Missing `TODO()`

File: test/org/lice/parse2/LexCommentTest.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+		Lexer l = new Lexer(src);
+		assertTrue(l.currentToken().getKind() == Token.TokenType.EOI);
+	}
+
+	@Test(timeout = 100)
+	public void testLexComment2() {
+		String src = "; Simply comments\n@dentifier";
+		Lexer l = new Lexer(src);
+		assertTrue(l.currentToken().getKind() == Token.TokenType.Identifier);
+		assertTrue(l.peekOneToken().getKind() == Token.TokenType.EOI);
```
Comment (ice1000) [inline test/org/lice/parse2/LexCommentTest.java]: Use `assertEquals`.

File: test/org/lice/parse2/LexIdentifierTest.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+	@Test(timeout = 100)
+	public void testLexIdentifiers2() {
+		String srcCode = "Fuck@dentifier,_Yet1A2not->her@dentifier";
+		Lexer l = new Lexer(srcCode);
+		assertTrue(l.currentToken().getKind() == Token.TokenType.Identifier);
+		assertTrue(l.currentToken().getStrValue().equals("Fuck@dentifier"));
+		l.nextToken();
+		assertTrue(l.currentToken().getKind() == Token.TokenType.Identifier);
+		assertTrue(l.currentToken().getStrValue().equals("_Yet1A2not->her@dentifier"));
+		assertTrue(l.peekOneToken().getKind() == Token.TokenType.EOI);
```
Comment (ice1000) [inline test/org/lice/parse2/LexIdentifierTest.java]: Use `assertEquals`.

**Review by ice1000 — COMMENTED** [review]
Review done.

File: src/org/lice/parse2/Lexer.java (hunk 5 above)
Comment (ice1000) [reply src/org/lice/parse2/Lexer.java]: Say, use `if (char in '0'..'9') {}`.

File: src/org/lice/parse2/Token.java (hunk 6 above)
Comment (chuigda) [reply src/org/lice/parse2/Token.java]: Fixed in https://github.com/lice-lang/lice-tiny/pull/1/commits/b67e48911be370300900bfbeabb22863a0dc3681 . It was I misunderstood the requirement at first.

File: src/org/lice/util/Error.kt (hunk 7 above)
Comment (chuigda) [reply src/org/lice/util/Error.kt]: implemented pretty-print https://github.com/lice-lang/lice-tiny/pull/1/commits/afba8143914b319bd0e5f2d74c5988416a182659

File: test/org/lice/parse2/LexIdentifierTest.java (hunk 9 above)
Comment (chuigda) [reply test/org/lice/parse2/LexIdentifierTest.java]: Switched to assertEquals in commit https://github.com/lice-lang/lice-tiny/pull/1/commits/a159cfb618965588b171b2d348c7e0a3e5ff68be .

**Comment by ice1000** [comment]
Looks nice.

Is that all done?

**Comment by chuigda** [comment]
I think we need more unit tests before converting this lexer into kotlin and merge it into master branch. Though the lexer performs well on single tokens, we still don't know whether it could solve a full program/expression/statement.

**Comment by ice1000** [comment]
Have you merged the newest progress of master branch?

**Comment by ice1000** [comment]
Try this, the generated stdlib:

```lisp
(-> false ()) ;; this is a stub, used for indexing
(-> null ()) ;; this is a stub, used for indexing
(-> true ()) ;; this is a stub, used for indexing
(defexpr ! ()) ;; this is a stub, used for indexing
(defexpr != ()) ;; this is a stub, used for indexing
(defexpr !== ()) ;; this is a stub, used for indexing
(defexpr % ()) ;; this is a stub, used for indexing
(defexpr & ()) ;; this is a stub, used for indexing
(defexpr && ()) ;; this is a stub, used for indexing
(defexpr * ()) ;; this is a stub, used for indexing
(defexpr + ()) ;; this is a stub, used for indexing
(defexpr - ()) ;; this is a stub, used for indexing
(defexpr -> ()) ;; this is a stub, used for indexing
(defexpr ->double ()) ;; this is a stub, used for indexing
(defexpr ->int ()) ;; this is a stub, used for indexing
(defexpr ->str ()) ;; this is a stub, used for indexing
(defexpr / ()) ;; this is a stub, used for indexing
(defexpr < ()) ;; this is a stub, used for indexing
(defexpr <= ()) ;; this is a stub, used for indexing
(defexpr == ()) ;; this is a stub, used for indexing
(defexpr === ()) ;; this is a stub, used for indexing
(defexpr > ()) ;; this is a stub, used for indexing
(defexpr >= ()) ;; this is a stub, used for indexing
(defexpr ^ ()) ;; this is a stub, used for indexing
(defexpr alias ()) ;; this is a stub, used for indexing
(defexpr def ()) ;; this is a stub, used for indexing
(defexpr def? ()) ;; this is a stub, used for indexing
(defexpr defexpr ()) ;; this is a stub, used for indexing
(defexpr deflazy ()) ;; this is a stub, used for indexing
(defexpr eval ()) ;; this is a stub, used for indexing
(defexpr exit ()) ;; this is a stub, used for indexing
(defexpr expr ()) ;; this is a stub, used for indexing
(defexpr extern ()) ;; this is a stub, used for indexing
(defexpr force|> ()) ;; this is a stub, used for indexing
(defexpr format ()) ;; this is a stub, used for indexing
(defexpr if ()) ;; this is a stub, used for indexing
(defexpr int->bin ()) ;; this is a stub, used for indexing
(defexpr int->hex ()) ;; this is a stub, used for indexing
(defexpr int->oct ()) ;; this is a stub, used for indexing
(defexpr lambda ()) ;; this is a stub, used for indexing
(defexpr lazy ()) ;; this is a stub, used for indexing
(defexpr load-file ()) ;; this is a stub, used for indexing
(defexpr print ()) ;; this is a stub, used for indexing
(defexpr rand ()) ;; this is a stub, used for indexing
(defexpr sin ()) ;; this is a stub, used for indexing
(defexpr sqrt ()) ;; this is a stub, used for indexing
(defexpr str->int ()) ;; this is a stub, used for indexing
(defexpr str->sym ()) ;; this is a stub, used for indexing
(defexpr str-con ()) ;; this is a stub, used for indexing
(defexpr sym->str ()) ;; this is a stub, used for indexing
(defexpr type ()) ;; this is a stub, used for indexing
(defexpr undef ()) ;; this is a stub, used for indexing
(defexpr when ()) ;; this is a stub, used for indexing
(defexpr while ()) ;; this is a stub, used for indexing
(defexpr | ()) ;; this is a stub, used for indexing
(defexpr |> ()) ;; this is a stub, used for indexing
(defexpr || ()) ;; this is a stub, used for indexing
(defexpr ~ ()) ;; this is a stub, used for indexing
```

File: src/org/lice/parse2/Token.java (hunk 6 above)
Comment (ice1000) [reply src/org/lice/parse2/Token.java]: Good!

File: src/org/lice/util/Error.kt (hunk 7 above)
Comment (ice1000) [reply src/org/lice/util/Error.kt]: Good!

File: test/org/lice/parse2/LexIdentifierTest.java (hunk 9 above)
Comment (ice1000) [reply test/org/lice/parse2/LexIdentifierTest.java]: Good!

**Comment by ice1000** [comment]
CI fix: due to `build.gradle` doesn't apply java plugin

**Comment by chuigda** [comment]
More "integrated" testcases have been added in commit https://github.com/lice-lang/lice-tiny/pull/1/commits/6829f98e37a4e9af26de7e669f1a66049e4c5a8a, and the test for "stdlib" also passed (I checked the result manually). Some small problems have also be solved now. I think it's time for us to migrate it into Kotlin.

**Comment by ice1000** [comment]
Ok. You do it first, and I'll review your changes.
In order to tell you some best practices, to help you do better during future contribution.

**Comment by ice1000** [comment]
`\f` is changed to `\u000C`.
Ref: https://www.compart.com/en/unicode/U+000C

**Comment by ice1000** [comment]
Don't worrry. AppVeyor is choking.

**Comment by chuigda** [comment]
I think it's time to replace the old `StringTree` in https://github.com/lice-lang/lice-tiny/blob/parser2/src/org/lice/model/Models.kt with something more advanced. We may use a `TokenTree` instead, since a `Token` carries more data than a `String` and may provide more accurate information in mapping and error reporting.

File: gradle/wrapper/gradle-wrapper.properties [hunk 10]
```diff
@@ -1,6 +1,6 @@
-#Fri Aug 11 17:18:00 CST 2017
+#Mon Jan 08 15:53:05 CST 2018
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-4.4-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.4-all.zip
```
Comment (ice1000) [inline gradle/wrapper/gradle-wrapper.properties]: Please revert this. @ILoveChenKX

**Comment by ice1000** [comment]
Reviewed. You've done a good job :+1:

File: test/org/lice/parse2/Sandbox.kt [hunk 11]
```diff
@@ -1,6 +1,8 @@
 package org.lice.parse2
 
 import org.junit.Test
+import org.lice.Lice
+import org.lice.parse.createRootNode
```
Comment (ice1000) [inline test/org/lice/parse2/Sandbox.kt]: Use Ctrl+Alt+O to remove unused imports.

**Pull request merged**